zoom-us: allow to select xdg-desktop-portal packages

Yarny0 opened the pull request:
## Things done

Implement what's been proposed in https://github.com/NixOS/nixpkgs/pull/397036#discussion_r2068124261 .  It wants a *function* that returns a list, not a simple list of packages.  Please see the commit message for details.

Note: I cannot really test this as I'm not using xdg-desktop-portal at all.

Notifying discussion participants: @philiptaron @gvolpe

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Yarny0 committed:
zoom-us: allow to select xdg-desktop-portal packages

To reduce the closure size of the zoom-us package,
this commit replaces the `xdgDesktopPortalSupport`
option with several new options that permit to (de-)select
several XDG desktop portal packages to be used by `zoom-us`.
Furthermore, two new options permit to inject further packages
(e.g. individual XDG desktop portal packages) into the virtual
filesystem that will be prepared by bubblewrap for zoom-us:

* `targetPkgs` is a *function* that maps the `pkgs`
  packages attrset to a list of packages.
* `targetPkgsFixed` is a simple list of packages.

`targetPkgs` should be preferred before
`targetPkgsFixed` if possible, as the `buildFHSEnv`
mechanism expects a function as explained above:
It needs to extract a list of dependency
packages for a `pkgs` package attrset.
Package injection likely also works if `targetPkgsFixed`
is used, but that creates a liability
in that changes of the `buildFHSEnv` implementation might
(at least more likely) break zoom-us in unexpected ways.

Note: This now *disables* all
XDG desktop portal packages by default.

Co-authored-by: Philip Taron <philip.taron@gmail.com>

gvolpe commented on pkgs/by-name/zo/zoom-us/package.nix:
```diff
@@ -3,7 +3,19 @@
   fetchurl,
   pulseaudioSupport ? true,
-  xdgDesktopPortalSupport ? true,
+  xdgDesktopPortalPkgs ? (
```
What do you think about having `xdg-desktop-portal` support disabled by default? I think most folks don't need it, e.g.

```
xdgDesktopPortalPkgs ? []
```

Then at use site you'd do this instead of adding a list of default packages:

```diff
-   ++ lib.optional (xdgDesktopPortalPkgs pkgs != [ ]) pkgs.xdg-desktop-portal
-   ++ xdgDesktopPortalPkgs pkgs;
+   ++ xdgDesktopPortalPkgs;
```

You mentioned the following [here](https://github.com/NixOS/nixpkgs/pull/397036#discussion_r2068757509):

> It's not that easy as buildFHSEnv wants a function that picks the right packages out of its sole pkgs argument. 

But I don't think this is necessarily the case. What I propose works just as fine, the override can be a simple list.

---

With this derivation in place, users that want `xdg-desktop-portal` support can easily override it, e.g.:

### Hyprland support

```nix
final: prev:

{
  zoom-us = prev.zoom-us.override {
    xdgDesktopPortalPkgs = [
      prev.xdg-desktop-portal;
      prev.xdg-desktop-portal-hyprland;
    ];
  };
}
```

Yarny0 replied:
> What do you think about having `xdg-desktop-portal` support disabled by default?

I don't know.  I would tend to keep the default values so that an ingenouos user gets as much functionality as possible.  A complaint that `zoom-us` uses too much disk space appears much less serious to me than a complaint that `zoom-us` does not allow screen sharing.  But my opinion in that case is very weak.  I'm not using `xdg-desktop-portal` myself, and I can't estimate how many users do.

> But I don't think this is necessarily the case. What I propose works just as fine, the override can be a simple list.

Well, it's probably not necessary, but I don't think the current interface is really bad either.  I try to avoid deviating from the `buildFHSEnv` API any more than needed, to reduce the probability of any future breakage (and I learnd that "it works" doesn't necessarily mean that everything works).

> With this derivation in place, users that want `xdg-desktop-portal` support can easily override it, e.g.:

With my current implementation, that's also quite simple:

```nix
final: prev:

{
  zoom-us = prev.zoom-us.override {
    xdgDesktopPortalPkgs = pkgs: [ pkgs.xdg-desktop-portal-hyprland; ];
  };
}
```

(Please tell me if I'm wrong; then that would be a bug in my implementation.)

If you really want to add a list, you can just do that:

```nix
final: prev:

{
  zoom-us = prev.zoom-us.override {
    xdgDesktopPortalPkgs = lib.const [ pkgs.xdg-desktop-portal-hyprland; ];
  };
}
```

That said, I just pushed two more commits that add a NixOS module (and a test).  That module adds the packages from `xdg.portal.extraPortals` without respecting the function's `pkgs` argument (like in the second example above).  I don't see an (easy) way of using that `pkgs` argument and the list in `xdg.portal.extraPortals` at the same time.  However, I try to at least keep the package itself as pure as possible.

gvolpe replied:
That works fine too 👍🏽

philiptaron commented:
I'm not convinced this is an upgrade yet.

Rather than having direct overrides be the main way this works and is embedded into someone's NixOS configuration, maybe it would be better to write a NixOS module for Zoom. That way, the sum total of the configuration could be taken into effect.

1. If `programs.hyprland.enable`, then this override
2. If `services.xserver.desktopManager.gnome.enable`, then a different override
3. etc

The best option would be one where the relevant tools are landed in the bwrap at _runtime_, rather than at system build time, based on `xdg.portal` and other configurations. Late binding, where the bwrap isn't aware of them during build, but if they're part of the system, makes use of them.

Yarny0 committed:
nixos/programs: add zoom-us module

The zoom-us package can be build with pulseaudio and with
support for one or more xdg-desktop-portal implementations.
This new module sets the proper build options
from the system configuration,
and adds the resulting package to `systemPackages`.
If an individual `zoom-us` package is assigned to the
`package` option with `package = zoom-us.override { ... };`,
arguments given to `.override` will take precedence
over the options determined by the module.
This permits to explicitly enable or disable some zoom-us
features while keeping the module's choices for other features.

Co-authored-by: Philip Taron <philip.taron@gmail.com>

Yarny0 committed:
nixos/tests: simple zoom-us test to verify the module

Yarny0 commented:
>  maybe it would be better to write a NixOS module for Zoom. That way, the sum total of the configuration could be taken into effect.

I added another commit which follows that way ... a bit.  I don't know much about hyperland or xdg-desktop-portals in general, but from looking at the `config/xdg/portal.nix` module, I grew the impression that using the `extraPortals` list seems a good idea.  That option's description even promises that GNOME and KDE add their respective portal implementations automatically; so with my second commit, they should also become a dependency of the zoom package.

I also added a commit with a very simple vm test that checks the module actually works.

Deal @philiptaron , please consider these commits as draft.  I'm not an xdg-desktop-portal user and I might totally misunserstand the `config/xdg/portal.nix` module.  The third commit especially got only authored as it was the easiest way for me to check if the new module works in general.  Feel free to change those commits as far as Github permits to do so, and/or to merge only a subset of my branch, as you see fit.

> The best option would be one where the relevant tools are landed in the bwrap at _runtime_

I'm not sure if I understand that correctly.  Where should the package look for the relevant tools?  Maybe in `/run/current-system/sw/...`?  To be honest, my instincts tell me to avoid having a package looking for libraries outside its own closure -- in particluar as the unmodified zoom binary is doing exactly that (it looks in `/usr/lib/...` for further libraries), and we should avoid that if possible.  If we go that road, we should at least ensure that the zoom package still works on non-NixOS systems where only Nix and nixpkgs is used.  But again, maybe I'm misunderstanding your idea here.

philiptaron commented:
This is a really interesting approach! I like the NixOS test. It'll be a little bit before I can dedicate the time to testing it out, but I would welcome feedback from others.

gvolpe commented:
@philiptaron I would suggest that we make a derivation that's easy to override, and NixOS (or Home Manager) modules can build on top of it, but I wouldn't make it mandatory to use NixOS modules to have the desired derivation.

I wouldn't want to rebuild my operating system just to install `zoom`. Home Manager is better suited to manage user applications for most folks, wouldn't want to impose a NixOS module. In any case, a HM module would make more sense.

Yarny0 commented:
> @philiptaron I would suggest that we make a derivation that's easy to override, and NixOS (or Home Manager) modules can build on top of it, but I wouldn't make it mandatory to use NixOS modules to have the desired derivation.
> 
> I wouldn't want to rebuild my operating system just to install `zoom`. Home Manager is better suited to manage user applications for most folks, wouldn't want to impose a NixOS module. In any case, a HM module would make more sense.

@gvolpe Can you try overriding the derivation like suggested in https://github.com/NixOS/nixpkgs/pull/403064#discussion_r2070618498 ?  I'm using zoom without the module, just with `environment.systemPackages = [ ( pkgs.zoom-us.override { xdgDesktopPortalPkgs = lib.const []; } ) ];`.  The module shouldn't be needed, it exists solely so that users don't have to enumerate their desktop portals twice.

Yarny0 commented:
Should we backport this to NixOS 25.05 ?

gvolpe commented:
Anything stopping this from being merged? @philiptaron

philiptaron commented:
This doesn't look right to me just yet -- in particular, the function-vs-list-of-packages difference doesn't seem right. Let me push that and let's take a look.

Yarny0 reviewed (commented):
Thanks @philiptaron .  This looks alright and works!  I have two suggestions, but I hesitate to update the branch.

What do you think about those suggestions?  I can implement them if you beckon me to do so.  I would also squash those changes into my original commits.

(it feels strange to review my own pull request :-) )

gvolpe commented on pkgs/by-name/zo/zoom-us/package.nix:
```diff
@@ -122,6 +160,20 @@ let
+          pkgs.xdg-desktop-portal-gnome
+          pkgs.xdg-desktop-portal-gtk
+        ])
+        ++ (lib.optionals hyprlandXdgDesktopPortalSupport [ pkgs.xdg-desktop-portal-hyprland ])
```
I don't love this. IIUC as a user, I wouldn't be able to override the `xdg-desktop-portal-hyprland` for `zoom` only? Maybe I'm the minority here, but I usually like to try out the latest Hyprland developments on my personal computer, but Zoom is needed for work so I try not to mess around with it.

philiptaron replied:
Actually, that's precisely what you are able to do!

1. Suppose you add `zoom-us` to `users.users.<username>.packages` or `environment.systemPackages`, as all have to do currently. That version, the default, won't bring in any XDG portal stuff, since the default for all of the `*XdgDesktopPortalSupport` is `false`. And any use of `override` will work: referencing `pkgs.zoom-us.override { hyprlandXdgDesktopPortalSupport = false; }` will guarantee that the XDG portal support isn't there.

2. Suppose you use the new-as-of-this-PR `programs.zoom-us.enable`. _By default_, that will integrate with your NixOS system, enabling and disabling the XDG desktop portal support that's appropriate for your system.

3. But we have a clever way to override the defaults: `programs.zoom-us.package`. Consider this: `programs.zoom-us.package = pkgs.zoom-us.override { hyprlandXdgDesktopPortalSupport = false; }` will ensure that even if `programs.hyprland.enable` is set, you'll continue to turn off that support in Zoom.

The logic also works in reverse. If you're enabling (say) Hyprland through modules which aren't in NixOS, and hence aren't checked by the logic above, you can always use `programs.zoom-us.package` to turn **on** features which are off by default.

philiptaron replied:
I'm in this same boat, so I futzed with it to make exactly this scenario work. Hence the `prev.hyprlandXdgDesktopPortalSupport  or ...` clauses in the module -- they're what enable your override to take effect.

philiptaron replied:
Ah, that's where the `targetPkgs` override comes in. Ignore all the `hyprlandXdgDesktopPortalSupport` business: that's there to power the NixOS module and those using NixOS to configure their systems.

```
  home.packages = [
    (zoom-us.override {
      targetPkgs = pkgs': [
        specific-hyprland-portal
        specific-anything-else
      ];
    })
  ];
```

The `pkgs'` is the `pkgs` passed to the function, which can used or not as your case demands.

There is obviously [overlays](https://wiki.nixos.org/wiki/Overlays), which would make `hyprlandXdgDesktopPortalSupport` work exactly as you want, but they're quite complicated.

philiptaron replied:
I kind of like it too; we could do this to remove the "function-ness".

```
  targetPkgs ? pkgs: extraPkgs,
  extraPkgs ? [],
```

philiptaron commented:
> What do you think about those suggestions? I can implement them if you beckon me to do so. I would also squash those changes into my original commits.

Absolutely. I pushed to your branch because I thought it was the lowest impact way of collaborating together, and because (as a committer) I had the rights to do so. My one request is to add Co-authored-by as commits get squashed together.

> (it feels strange to review my own pull request :-) )

😀

Yarny0 commented:
Updated the branch

* backward-rebased onto [current `nixos-unstable`](https://github.com/NixOS/nixpkgs/commit/2795c506fe8fb7b03c36ccb51f75b6df0ab2553f) -- this avoids a lot of package rebuilds when testing
* squashed [@philiptaron's commit](https://github.com/NixOS/nixpkgs/pull/403064/commits/a5cd70c1a332c63a48858e72aa6b5912416fef34) into the package and module commit
* added the [idea of a second packages argument](https://github.com/NixOS/nixpkgs/pull/403064#discussion_r2103111989)

Some remarks on the code:

* In contrast to [@philiptaron's commit](https://github.com/NixOS/nixpkgs/pull/403064/commits/a5cd70c1a332c63a48858e72aa6b5912416fef34), I have rearranged the `linuxGetDependencies` function as to avoid the `let .. in`.
* I renamed the `extraPkgs` to `targetPkgsFixed` to stick to the original name `targetPkgs` of the `buildFHSEnv` mechanism.

I hope this is OK.

After backporting the pull request locally onto [current `nixos-24.11`](https://github.com/NixOS/nixpkgs/commit/2baa12ff69913392faf0ace833bc54bba297ea95), I tested some of these items in Plasma5/X11/Pulseaudio (unchecked means "not tested"):

- [x] sidebar loads (albeit empty on my installation, cf. https://github.com/NixOS/nixpkgs/issues/344931 )
- [x] audio is "detected" by zoom: microphone VU-meter shows amplitude (not tested in VirtualBox)
- [x] audio is produced: "Test speaker" generates audible ting-a-ling (not tested in VirtualBox)
- [ ] other participant's audio in video conference is audible (not tested in VirtualBox)
- [ ] other participants can hear me
- [x] sending camera picture in video conference works
- [ ] participant's camera video is received in video conference
- [x] screen sharing sends screen to other participants
- [ ] participant's shared screen is displayed
- [x] "Participants" dialog box is shown (cf. https://github.com/NixOS/nixpkgs/issues/116355 )
- [x] reading team chat messages works as expected
- [x] settings dialog is shown/useable
- [x] SSO works: Zoom starts Firefox, then Firefox calls back to Zoom after login, as usual
- [ ] writing team chat messages
- [ ] sidebar shows content
- [ ] Pipewire
- [ ] Wayland
- [ ] xdg-desktop-portal (cf. https://github.com/NixOS/nixpkgs/issues/359533 )
- [ ] Darwin

> I pushed to your branch because I thought it was the lowest impact way of collaborating together, and because (as a committer) I had the rights to do so.

That's very much OK -- I'm just not used to that and therefore careful.

> My one request is to add Co-authored-by as commits get squashed together.

Thanks for that hint.  I have to admit I wouldn't have thought about that, without any ill-meaning.  And please have a look at the credits in the commit message and tell me if something is not OK.

Note: Zoom released a new version.  A superficial test caused problems (the settings dialog won't show up).  I will investigate and maybe open a separate PR for the update.  *sigh*

philiptaron commented on pkgs/by-name/zo/zoom-us/package.nix:
```diff
@@ -179,17 +222,19 @@ let
+    ++ lib.optional hyprlandXdgDesktopPortalSupport pkgs.xdg-desktop-portal-hyprland
+    ++ lib.optional wlrXdgDesktopPortalSupport pkgs.xdg-desktop-portal-wlr
+    ++ lib.optional xappXdgDesktopPortalSupport pkgs.xdg-desktop-portal-xapp
+    ++ targetPkgs pkgs
+    ++ targetPkgsFixed;
```
This double-adds packages, since by default `targetPkgs` returns whatever's in `targetPkgsFixed`. Suggest making `targetPkgs` return back to `pkgs: []`.

philiptaron reviewed (changes requested):
One change requested.
